contracts/token/ERC20/extensions/ERC1363.sol

```diff
@@ -0,0 +1,174 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.20;
+
+import {ERC20} from "../ERC20.sol";
+import {IERC165, ERC165} from "../../../utils/introspection/ERC165.sol";
+
+import {IERC1363} from "../../../interfaces/IERC1363.sol";
+import {IERC1363Receiver} from "../../../interfaces/IERC1363Receiver.sol";
+import {IERC1363Spender} from "../../../interfaces/IERC1363Spender.sol";
+
+/**
+ * @title ERC1363
+ * @dev Extension of {ERC20} tokens that adds support for code execution after transfers and approvals
+ * on recipient contracts. Calls after transfers are enabled through the {ERC1363-transferAndCall} and
+ * {ERC1363-transferFromAndCall} methods while calls after approvals can be made with {ERC1363-approveAndCall}
+ */
+abstract contract ERC1363 is ERC20, ERC165, IERC1363 {
+    /**
+     * @dev Indicates a failure with the token `receiver`. Used in transfers.
+     * @param receiver Address to which tokens are being transferred.
+     */
+    error ERC1363InvalidReceiver(address receiver);
+
+    /**
+     * @dev Indicates a failure with the token `spender`. Used in approvals.
+     * @param spender Address that may be allowed to operate on tokens without being their owner.
+     */
+    error ERC1363InvalidSpender(address spender);
+
+    /**
+     * @inheritdoc IERC165
+     */
+    function supportsInterface(bytes4 interfaceId) public view virtual override(ERC165, IERC165) returns (bool) {
+        return interfaceId == type(IERC1363).interfaceId || super.supportsInterface(interfaceId);
+    }
+
+    /**
+     * @dev Moves a `value` amount of tokens from the caller's account to `to`
+     * and then calls {IERC1363Receiver-onTransferReceived} on `to`.
+     *
+     * NOTE: Reverts if `to` is an account without code or the recipient contract does not implement
+     * {IERC1363Receiver-onTransferReceived}
+     */
+    function transferAndCall(address to, uint256 value) public returns (bool) {
+        return transferAndCall(to, value, "");
+    }
+
+    /**
+     * @dev Moves a `value` amount of tokens from the caller's account to `to`
+     * and then calls {IERC1363Receiver-onTransferReceived} on `to`.
+     *
+     * NOTE: Reverts if `to` is an account without code or the recipient contract does not implement
+     * {IERC1363Receiver-onTransferReceived}
+     */
+    function transferAndCall(address to, uint256 value, bytes memory data) public virtual returns (bool) {
+        transfer(to, value);
```

Comment:
A potential issue with using the ERC20 `transfer` function like this is that someone can override `transfer` and return `false`. It'd be within the ERC20 spec but it'll allow calls to arbitrary `to`. An ERC1363Receiver shouldn't blindly trusts calls from tokens since they may be triggered unconditionally. 

I'm thinking we shoud:

- Wrap the `transfer` function in an internal `function _callTransfer(address to, uint256 value) private returns (bool)` so that users can hook to this return value using `super` and decide what to do with it (see next point).
- Add a warning in the `*AndCall` functions indicating that the function doesn't not check the return value of the underlying `approve`/`transfer` and users MUST override `_callTransfer` to adapt for ERC20's returning false.

Reply:
I would really avoid adding more internal / private functions. There are already a lot. 

We could do 
```solidity
    function transferAndCall(address to, uint256 value, bytes memory data) public virtual returns (bool) {
        if (!transfer(to, value)) {
            revert ERC1363TransferFailed(to, value);
        }
        _checkOnTransferReceived(_msgSender(), to, value, data);
        return true;
    }
    ```

Reply:
Right, this makes sense to avoid wrapping `transfer` in an internal `_callTransfer`. I'm ok with it.